Multigpu

fral92 opened the pull request:
Fix for multigpu issue

fral92 committed:
Fix multi-gpu fault in the case of smaller batches

* tf.split removed from graph, the placeholders are now in a list
* split the operations of the graph according to the number of working towers
* modify compute_chunk_size function in utils to compute the chunks directly

fral92 committed:
Fix multi-gpu fault in the case of smaller batches

* tf.split removed from graph, the placeholders are now in a list
* split the operations of the graph according to the number of working towers
* modify compute_chunk_size function in utils to compute the chunks directly

fvisin commented:
@fral92 can you rebase on master please? I'll check the PR in the meanwhile.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,30 @@ def __run(build_model):
+
+        # Create input placeholders for each gpu tower
+        # All the placeholders are instantiated with a default value
+        # since tensorflow result on a fault if one placeholder is not
```
since we want to be able to use a subset of them when the batchsize is not big enough to fill all the GPUs and TF raises a negative shape error if one placeholder is not fed (and has no default value). This is not a problem since when the default values is being used we ignore it's output and do not perform gradient descent on it.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,30 @@ def __run(build_model):
+                shape=cfg.input_shape))
+            val_inputs_per_gpu.append(tf.placeholder_with_default(
+                np.zeros(shape=[1]+cfg.input_shape[1:], dtype=cfg._FLOATX),
+                shape=cfg.input_shape))
```
Is `cfg.input_shape` correct here? If so, why do you need two different placeholders?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,30 @@ def __run(build_model):
+                l_shape = np.prod(cfg.input_shape[2:])
+            else:
+                l_shape = np.prod(cfg.input_shape[1:])
+            labels_per_gpu.append(tf.placeholder_with_default(
```
Add a comment saying that shape is None cause we provide the labels flattened.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        # The actual number of gpus used for the smaller batch
+        # for each set ('train', 'valid', 'test') is saved in
+        # a dictionary
+        smaller_n_gpus = {}
```
I'd call it `n_gpus_per_set`

Change the comment in:
When the set size is not a multiple of the batch_size, the last batch will be smaller than batch_size. When this happen we might not be able to feed all the GPUs with an input for this batch. Here we compute the number of GPUs to use for the last batch of each set.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        # a dictionary
+        smaller_n_gpus = {}
+        # Training case
+        d_params = copy.deepcopy(cfg.dataset_params)
```
Rename to `train_dataset_params`.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        nbatches = train.nbatches
+        batch_size = cfg.dataset_params['batch_size']
+        smaller_batch = (nsamples -
+                         (nbatches - 1) * d_params['batch_size'])
```
You can use `batch_size` rather than `d_params['batch_size']` here, which should allow you to use a better indentation hopefully.

fvisin replied:
Also, you might be able to use train.nsamples and train.nbatches directly and save three lines (check if indentation is doable in this case).

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        nsamples = train.nsamples
+        nbatches = train.nbatches
+        batch_size = cfg.dataset_params['batch_size']
+        smaller_batch = (nsamples -
```
Call this `last_batch_size`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        batch_size = cfg.dataset_params['batch_size']
+        smaller_batch = (nsamples -
+                         (nbatches - 1) * d_params['batch_size'])
+        s_n_gpus_train = int(smaller_batch/batch_size)
```
- `smaller_batch/batch_size` should already be an int (in python 2.7) since both arguments should be ints if I'm not wrong. To be python3 compatible, you can use the integer division (`smaller_batch // batch_size`).
- add a space between the `/` (or `//`) operator
- either find a better name (e.g., `last_batch_n_gpus`) or put it directly in the dictionary

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+                         (nbatches - 1) * d_params['batch_size'])
+        s_n_gpus_train = int(smaller_batch/batch_size)
+        rem = smaller_batch % batch_size
+        if rem != 0:
```
Do not create the `rem` variable, merge this line and the previous one please.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        # s_num_pixels = {}
+        for s in cfg.val_on_sets:
+            v_params = copy.deepcopy(cfg.valid_params)
+            v_params['batch_size'] *= cfg.num_splits
```
It looks like `v_params` is exactly the same as `d_params`. Do you actually need two of them?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+                            **d_params)
+        nsamples = train.nsamples
+        nbatches = train.nbatches
+        batch_size = cfg.dataset_params['batch_size']
```
If I am not wrong, if you do this assignment before multiplying batch_size by  cfg.num_splits (https://github.com/fvisin/main_loop_tf/pull/12/files#diff-7fb60c0a94ff78e23dd1187630e2a7bbR337) you can loose the copy of the dataset params.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +323,46 @@ def __run(build_model):
+        smaller_n_gpus['train'] = s_n_gpus_train
+        # Validation case
+        # s_num_pixels = {}
+        for s in cfg.val_on_sets:
```
Code repetition (Software engineering 101)!! :P 
Move the 'train' case you computed so far in this loop by adding it to the val_on_sets list.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -324,11 +371,13 @@ def __run(build_model):
-            val_outs, val_summary_ops, val_reset_cm_op = build_graph(
-                val_placeholders, cfg.val_input_shape, cfg.Optimizer,
-                cfg.weight_decay, cfg.loss_fn, build_model, False)
+            for s in cfg.val_on_sets:
```
`val_outs`, `val_summary_ops`, `val_reset_cm_op` will be overwritten at each iteration. Is this the intended behavior? Since the number of gpus changes, I guess you'd rather want them to be dictionaries and have one for each `s`.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -417,26 +467,43 @@ def __run(build_model):
+# in the name.
+# TODO: double check if this way to operate is correct.
+# In the specific check if duplicating the operation could affect in
+# in some manner the learning task.
```
Next time comment in the PR rather than in the code, so that it's easier to keep trace of the discussion! ;)

fvisin replied:
Move this comment inside the function, as a docstring and change it (with proper indentation) to
```python
''' Build the multiGPU graph of computation

This function creates a copy of the computation graph on each GPU. The result of the computation of each GPU is stored in a "tower"
Note that thanks to the use of name_scopes and variable_scopes, calling this function multiple times does not create multiple copies of the *Ops* and of the *Variables* (respectively), but rather only adds the Ops that change from one call to the other and reuses the same Variables.

Furthermore, to accommodate for the case where the last minibatch of a set is smaller than batch_size, we create two versions of each operation (e.g., those that compute the summaries), the first that uses all the GPUs (default) and the second that only uses the subset that is possible to use in the reduced batch case. Each returned value is then a list of two ops. This allows to choose at runtime which operations of the graph to call, depending on the batch size.
'''
```

fvisin commented on main_loop_tf/main.py:
```diff
@@ -417,26 +467,43 @@ def __run(build_model):
-    for gpu_input in inputs_per_gpu:
-        gpu_input.set_shape(input_shape)
+    labels = tf.concat(labels_per_gpu, axis=0)
+    smaller_labels = tf.concat(labels_per_gpu[:smaller_n_gpus], axis=0)
```
Move these two where you use them (i.e., after the long loop)!

fvisin replied:
Be consistent with the names. If you go for `s_something` use this convention everywhere. `smaller_labels` --> `s_labels`. Also add a comment before the definition of `labels`: concatenate the per-gpu placeholders to get a placeholder for the full (f) list of gpus and one for the subset (s) to be used for the minibatch with less batches.

fvisin replied:
Also rename labels --> f_labels (or better labels_f)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -576,27 +643,48 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
     # Compute the mean IoU
     # TODO would it be better to use less precision here?
     mask = tf.ones_like(labels)
+    s_mask = tf.ones_like(smaller_labels)
```
Replace this section with a for loop like:
```python
for labels, preds, softmax_preds in zip([f_labels, s_labels], [f_preds, s_preds], [f_softmax_preds, s_softmax_preds]):
    mask = tf.ones_like(labels)
    if len(cfg.void_labels):
        mask = tf.cast(tf.less_equal(labels, nclasses), tf.int32)
    preds_flat = tf.reshape(preds, [-1])
...
```

fvisin replied:
I suggest you also consider the case where the number of samples is a multiple of batch_size (i.e. where we don't need two graphs). You can do this before the for loop (and zip on out_list). 
```python
if n_small_gpus:
    out_vars = ([f_f_labels, s_labels], [f_preds, s_preds], [f_softmax_preds, s_softmax_preds])
else:
    out_vars = ([f_f_labels], [f_preds], [f_softmax_preds])
```

There might be a better name than out_vars for this. Pick the one you prefer.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -576,27 +643,48 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
     preds = tf.concat(tower_preds, axis=0)
+    s_preds = tf.concat(tower_preds[:smaller_n_gpus], axis=0)
     softmax_preds = tf.concat(tower_soft_preds, axis=0)
+    s_softmax_preds = tf.concat(tower_soft_preds[:smaller_n_gpus],
```
Add the `f` prefix (or suffix) to preds and softmax_preds as well.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -676,6 +795,10 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
 
+    # Compute smaller batch if any
+    smaller_batch_size = (train.nsamples -
+                          (train.nbatches - 1) * dataset_params['batch_size'])
```
You don't need this. Just check for batch_size rather than for smaller_batch_size

fvisin commented on main_loop_tf/main.py:
```diff
@@ -715,12 +836,22 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
-                x_batch.shape[0], np.prod(train.data_shape[:2]))
+            # Compute the actual number of gpus used
+            # for the current batch
+            if current_size == smaller_batch_size:
```
Ditto.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -715,12 +836,22 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
+
+            # The function now directly split the inputs and labels
+            # according to the actual number of gpus used
+            # (tf.split in build graph has been removed)
```
Remove the comment

fvisin commented on main_loop_tf/main.py:
```diff
@@ -715,12 +836,22 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
+            # The function now directly split the inputs and labels
+            # according to the actual number of gpus used
+            # (tf.split in build graph has been removed)
+            x_batch_chunks, y_batch_chunks = compute_chunks(x_batch,
```
Rename it to `split_in_gpu_chunks`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -730,19 +861,27 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
-                         1 + loss_value]
-            feed_dict = {p: v for (p, v) in zip(placeholders, in_values)}
+            in_values = x_batch_chunks + y_batch_chunks + [loss_value]
+            if current_size == smaller_batch_size:
```
ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -730,19 +861,27 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
             # train_op does not return anything, but must be in the
             # outputs to update the gradient
             if cum_iter % cfg.train_summary_freq == 0:
+                t_outs = train_outs[0] + [train_summary_op[0]]
```
Put this in the else, as usual

fvisin commented on main_loop_tf/main.py:
```diff
@@ -730,19 +861,27 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
                 sv.summary_computed(cfg.sess, summary_str)
             else:
-                loss_value, _ = cfg.sess.run(train_outs, feed_dict=feed_dict)
+                t_outs = train_outs[0]
```
Ditto

fvisin commented on main_loop_tf/utils.py:
```diff
@@ -21,26 +21,25 @@
+# function that I found to be quite good in splitting the
+# batch equally among the gpus (with the previous method
+# the splits were almost unbalanced)
+def compute_chunks(x_batch, y_batch, gpus_used):
```
Ditto (rename and remove comment)

fvisin commented on main_loop_tf/validate.py:
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -75,6 +85,16 @@ def validate(placeholders,
+        current_size = x_batch.shape[0]
+        # Compute the actual number of gpus used
+        # for the current batch
+        if current_size == smaller_batch_size:
```
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -75,6 +85,16 @@ def validate(placeholders,
         raw_data_batch = ret['raw_data']
 
+        # Current batch size
+        current_size = x_batch.shape[0]
```
It's almost as long....get rid of the variable or use a shorter name

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -75,6 +85,16 @@ def validate(placeholders,
+        # Compute the actual number of gpus used
+        # for the current batch
+        if current_size == smaller_batch_size:
+            n_gpus_used = smaller_n_gpus[which_set]
```
Rename to this_n_gpus

fvisin commented on main_loop_tf/main.py:
```diff
@@ -715,12 +836,22 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
+            # Compute the actual number of gpus used
+            # for the current batch
+            if current_size == smaller_batch_size:
+                n_gpus_used = smaller_n_gpus['train']
```
Rename to this_n_gpus

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -103,11 +123,15 @@ def validate(placeholders,
             x_in = x_batch
             y_in = y_batch
 
+        x_in_chunks, y_in_chunks = compute_chunks(x_in, y_in, n_gpus_used)
```
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -12,16 +12,22 @@
              which_set='valid',
              epoch_id=None):
 
+    val_inputs_per_gpu, labels_per_gpu = placeholders
```
Loose these variables. You use them in the next 2 lines only, and the names are longer than placeholders[i]...

fvisin commented on main_loop_tf/main.py:
```diff
@@ -634,22 +722,53 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+    # the case of smaller batch)
+    p_list = inputs_per_gpu + labels_per_gpu + [prev_err]
+    smaller_p_list = (inputs_per_gpu[:smaller_n_gpus['train']] +
+                      labels_per_gpu[:smaller_n_gpus['train']] + [prev_err])
```
Rather than saving two p_lists, why don't you create `p_list` in the main loop depending on the size of the batch (i.e., either equal to the full sum or the sum of the subsets) and just use it as is later on?

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -12,16 +12,22 @@
+    val_inputs_per_gpu, labels_per_gpu = placeholders
+    p_list = val_inputs_per_gpu + labels_per_gpu
+    smaller_p_list = (val_inputs_per_gpu[:smaller_n_gpus[which_set]] +
+                      labels_per_gpu[:smaller_n_gpus[which_set]])
```
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -103,11 +123,15 @@ def validate(placeholders,
+        if current_size == smaller_batch_size:
+            feed_dict = {p: v for (p, v) in zip(smaller_p_list, in_values)}
+        else:
+            feed_dict = {p: v for (p, v) in zip(p_list, in_values)}
```
Ditto

fvisin commented on main_loop_tf/validate.py:
```diff
@@ -121,14 +145,20 @@ def validate(placeholders,
             # Get batch pred, mIoU so far, batch loss and potentially the
             # summary
             if cidx % cfg.val_summary_freq == 0:
+                v_outs = eval_outs[0] + [val_summary_op[0]]
```
You can compute v_outs outside of the summary_freq if and then use v_outs[0] in the else branch at L 157

fvisin replied:
Actually if you do it before the set_has_GT if you can then reuse it also in the non set_has_GT branch at L 177

fvisin reviewed (changes requested):
There are some things that can be improved, but overall it seems to be ok. Please implement the changes I asked and push a new commit AND, after that, check that it works with validation as well.

fral92 committed:
Refactoring multigpu PR

fral92 committed:
Refactoring multigpu PR

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,34 @@ def __run(build_model):
+        # Create input placeholders for each gpu tower
+        # All the placeholders are instantiated with a default value
+        # since we want to be able to use a subset of them when the
+        # batchsize is not big enough to fill all th GPUs and TF raises
```
*the

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,34 @@ def __run(build_model):
+        # This is not a problem since when the default value is being
+        # used we ignore it's output and do not perform gradient descent
+        # on it.
+        # Note: shape for labels placeholders will be None since we
```
Move the comment near the corresponding line of code (i.e., where my comment was)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,34 @@ def __run(build_model):
+                np.zeros(shape=[1]+cfg.input_shape[1:], dtype=cfg._FLOATX),
+                shape=cfg.input_shape))
+            val_inputs_per_gpu.append(tf.placeholder_with_default(
+                np.zeros(shape=[1]+cfg.val_input_shape[1:], dtype=cfg._FLOATX),
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -258,11 +259,34 @@ def __run(build_model):
+        labels_per_gpu = []
+        for _ in range(cfg.num_splits):
+            inputs_per_gpu.append(tf.placeholder_with_default(
+                np.zeros(shape=[1]+cfg.input_shape[1:], dtype=cfg._FLOATX),
```
Why do you set the batchsize to 1?

fvisin replied:
@fral92 can you comment on this please?

fvisin replied:
Nevermind, I decided to use `batch_size` and `val_batch_size`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +327,30 @@ def __run(build_model):
+        val_placeholders = [val_inputs_per_gpu, labels_per_gpu]
+
+        # When the set size is not a multiple of the batch_size, the
+        # last batch will be smaller than batch_size. When this happen
```
happens

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +327,30 @@ def __run(build_model):
+            if s == 'train':
+                params = cfg.dataset_params
+            else:
+                params = cfg.valid_params
```
You might want to use a oneliner here if it doesn't overflow in the next line:
`params = cfg.dataset_params if s == 'train' else cfg.valid_params`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +327,30 @@ def __run(build_model):
+            else:
+                params = cfg.valid_params
+            d_params = copy.deepcopy(params)
+            d_params['batch_size'] *= cfg.num_splits
```
Add a comment before L345: Request the dataset one minibatch \*per gpu\*

fvisin commented on main_loop_tf/main.py:
```diff
@@ -303,19 +327,30 @@ def __run(build_model):
+            last_batch_size = (this_set.nsamples -
+                               (this_set.nbatches - 1) *
+                               batch_size * cfg.num_splits)
+            n_gpus_per_set[s] = last_batch_size // batch_size
```
Add a comment: spread the last batch over the GPUs as evenly as possible

fvisin commented on main_loop_tf/main.py:
```diff
@@ -324,11 +359,18 @@ def __run(build_model):
                 placeholders, cfg.input_shape, cfg.Optimizer, cfg.weight_decay,
-                cfg.loss_fn, build_model, True)
+                cfg.loss_fn, build_model, True, n_gpus_per_set['train'],
+                'train')
```
Please remove cfg.Optimizer, cfg.weight_decay and cfg.loss_fn (as they don't differ in training and validation and they can be recovered from cfg)

Remove it also from L370 and from the `build_graph` definition. Possibly do this in a separate commit so that it doesn't get mixed with the multiGpu stuff.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -324,11 +359,18 @@ def __run(build_model):
+                 val_reset_cm_op[s]) = build_graph(
+                     val_placeholders, cfg.val_input_shape,
+                     cfg.Optimizer, cfg.weight_decay, cfg.loss_fn,
+                     build_model, False, n_gpus_per_set[s], s)
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -324,11 +359,18 @@ def __run(build_model):
+            val_outs = {}
+            val_summary_ops = {}
+            val_reset_cm_op = {}
+            for s in cfg.val_on_sets:
```
[Minor comment, feel free to ignore]
You could also iterate over a zip, which is more explicit / pythonic:
`for s, val_out, val_summary_op in zip(cfg.val_on_sets, val_outs, val_summary_ops)`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -414,41 +457,53 @@ def __run(build_model):
 
 def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                build_model, is_training):
+                build_model, is_training, s_gpus, which_set):
```
Ditto

fvisin replied:
Rename `s_gpus` as either `s_n_gpus` or `this_n_gpus`

fvisin replied:
You can loose is_training and check if which_set is == 'train' instead.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -414,41 +457,53 @@ def __run(build_model):
     tower_soft_preds = []
     tower_losses = []
-    summaries = {}
+    # summaries = {}
```
Remove this

fvisin commented on main_loop_tf/main.py:
```diff
@@ -479,8 +534,7 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
                     # Save this GPU's loss summary
-                    for k, s in summaries.iteritems():
-                        s.append(tf.summary.scalar('Loss', loss))
+                    summaries.append(tf.summary.scalar('Loss', loss))
```
I wonder if we should add `which_set` to the name scope. Comments on this?

fvisin replied:
@marcociccone @fral92 any comment on this?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -514,7 +568,7 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
 
-                            summaries["training"].append(tf.summary.scalar(
+                            summaries.append(tf.summary.scalar(
                                 "Tower%d_NoiseGrad" % dev_idx,
```
We don't need the "tower", right? It's in the namescope already.

fvisin replied:
Check why 'GPU' does not show up in front of the name in Tensorboard

fvisin commented on main_loop_tf/main.py:
```diff
@@ -525,7 +579,7 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
 
-                            summaries["training"].append(tf.summary.scalar(
+                            summaries.append(tf.summary.scalar(
                                 "Tower%d_NoiseGrad" % dev_idx,
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -551,20 +605,17 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                                                 (dev_idx, var_name),
-                                                 grad_values))
+                        summaries.append(tf.summary.histogram(
+                            "Tower%d_Gradients_%s" % (dev_idx, var_name),
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -551,20 +605,17 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                                              (dev_idx, var_name),
-                                              tf.global_norm([grad_values])))
+                        summaries.append(tf.summary.scalar(
+                            "Tower%d_GradientNorm_%s" % (dev_idx, var_name),
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -551,20 +605,17 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                                      dev_idx,
-                                      tf.global_norm(list(zip(*grads))[0])))
+                summaries.append(tf.summary.scalar(
+                    "Tower%d_Global_norm/clipped_grad_norm" % dev_idx,
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -324,11 +359,18 @@ def __run(build_model):
+                 val_reset_cm_op[s]) = build_graph(
+                     val_placeholders, cfg.val_input_shape,
+                     cfg.Optimizer, cfg.weight_decay, cfg.loss_fn,
+                     build_model, False, n_gpus_per_set[s], s)
```
replace `s` with `'val_' + s`

fvisin replied:
You don't need "is_training" anymore as well.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -414,41 +457,53 @@ def __run(build_model):
-            summaries[k] = tf.get_collection_ref(key='val_' + k + '_summaries')
+        summaries = tf.get_collection_ref(
+            key='val_' + which_set + '_summaries')
     tower_suffix = 'train' if is_training else 'val'
```
remove `tower_suffix` and replace with `which_set` in the rest of the function

fvisin commented on main_loop_tf/main.py:
```diff
@@ -525,7 +579,7 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                            summaries["training"].append(tf.summary.scalar(
+                            summaries.append(tf.summary.scalar(
                                 "Tower%d_NoiseGrad" % dev_idx,
                                 grad_noise_scale))
```
@marcociccone is there a reason why we have this summary once again (see L571)? It looks like a copy and paste that has not been modified.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -574,29 +625,54 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+    (preds, labels, softmax_preds, mask, preds_flat,
+     avg_tower_loss, grads_and_vars,
+     train_op, m_iou, per_class_iou, cm_update_op,
+     reset_cm_op) = ([None] * 2 for _ in range(12))
```
Looks like you could put more variables on the same line and squeeze this. Check it please.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -574,29 +625,54 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+     train_op, m_iou, per_class_iou, cm_update_op,
+     reset_cm_op) = ([None] * 2 for _ in range(12))
+    update_ops = [[], []]
+    for idx, n_gpus in enumerate(num_gpus):
```
All these lists could be dictionaries with e.g., 'full_size' and 'small' or 'small_size' as keys. The `idx` could become e.g., `batchsize_case` or `n_gpus_case`. Feel free to propose a better name. It would be easier to understand what's happening (rather than relying on the unspoken rule that the first element of the array is the full case and the second is the small batch case) later on.

I leave it up to you to decide whether this is necessary or not, but at least rename idx --> case or something similar.

fvisin replied:
Do not use `num_gpus` and `n_gpus`, it's very confusing. Rename to e.g., `n_gpus_per_case`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -574,29 +625,54 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+            mask[idx] = tf.cast(tf.less_equal(labels[idx], nclasses), tf.int32)
+        preds_flat[idx] = tf.reshape(preds[idx], [-1])
+
+        (m_iou[idx], per_class_iou[idx],
```
Weird line break again. Break it after `cm_update_op[idx],`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -574,29 +625,54 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+
+        if is_training:
+            # Impose graph dependency so that update operations are computed
+            # even if they're are not explicit in the outputs os session.run
```
*of

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+        # Duplicate summaries: this allow to distiguish between the
+        # avg_tower_loss in the case of full num of gpus and that in
+        # the case of a subset of num of gpus
+        list_summaries = [summaries, summaries[:]]
```
Why is this `list_summaries = [summaries, summaries[:]]` rather than `list_summaries = [summaries, summaries]` ?
Also, it's not clear to me why would you need a list.

fvisin replied:
In fact, you don't! Remove this line.

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+        # the case of a subset of num of gpus
+        list_summaries = [summaries, summaries[:]]
+        train_summary_op = [[], []]
+        val_summary_op = [[], []]
```
This is so ugly. Why would you need this? Can't you just use append?

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+        list_summaries = [summaries, summaries[:]]
+        train_summary_op = [[], []]
+        val_summary_op = [[], []]
+        for idx, n_gpus in enumerate(num_gpus):
```
Once again idx could be the "ngpu_case"

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+
+            # Scalars
+            # list_summaries[idx].append(tf.summary.scalar(
+            #     'Mean_tower_loss_' + which_set, avg_tower_loss[idx]))
```
Remove the commented lines unless there is a reason to keep them

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
 
-        # During the training we want to save informations about the
-        # gradients, the trainable variables and the activations.
+            # During the training we want to save informations about the
```
the training --> training

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-            for k, s in summaries.iteritems():
-                val_summary_ops[k] = tf.summary.merge(s)
+            if is_training:
+                # Merge only the training summaries of the gpus used.
```
that are being used

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                        #         'Mean_tower_loss_' + which_set in el.name):
+                        #     training_summaries += [el]
+                train_summary_op[idx] = tf.summary.merge(training_summaries)
+                import ipdb; ipdb.set_trace()
```
!!!

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                # This is done searching by name since the variables are
+                # always preceded by 'Tower{0,1,2,..}' or 'GPU{0,1,2,..}'
+                training_summaries = []
+                for i in range(n_gpus):
```
use a more comprehensible variable name, e.g., n_gpu

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                for i in range(n_gpus):
+                    for el in list_summaries[idx]:
+                        if ('GPU%d' % i in el.name or
+                                'Tower%d' % i in el.name):
```
Remove the Tower as we will not have it anymore. Also remove from comment L722

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                # always preceded by 'Tower{0,1,2,..}' or 'GPU{0,1,2,..}'
+                training_summaries = []
+                for i in range(n_gpus):
+                    for el in list_summaries[idx]:
```
you don't need a list. Just replace `list_summaries[idx]` with `summaries`

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-        return [avg_tower_loss, train_op], train_summary_op, reset_cm_op
+        return ([[avg_tower_loss[0], train_op[0]],
+                 [avg_tower_loss[1], train_op[1]]],
+                train_summary_op, reset_cm_op)
```
You can move the return to L717 (and the same for the non is_training case)

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                # above)
+                val_summaries = []
+                for i in range(n_gpus):
+                    for el in list_summaries[idx]:
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -618,37 +694,72 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                for i in range(n_gpus):
+                    for el in list_summaries[idx]:
+                        if ('GPU%d' % i in el.name or
+                            'Tower%d' % i in el.name):
```
Ditto

fvisin commented on main_loop_tf/main.py:
```diff
@@ -715,12 +824,16 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
-            # TODO evaluate if it's possible to pass num_splits inputs in
-            # a list, rather than the input as a whole and the shape of
-            # the splits as a tensor.
+            # Compute the actual number of gpus used
```
Line too short

fvisin commented on main_loop_tf/main.py:
```diff
@@ -730,19 +843,39 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
+                                  placeholders[1][:s_gpus['train']] +
+                                  [placeholders[2]])
+                feed_dict = {p: v for(p, v) in zip(s_placeholders,
+                                                   in_values)}
```
MINOR: This can be moved inside the previous if/else on the same condition to keep similar things together

fvisin commented on main_loop_tf/main.py:
```diff
@@ -730,19 +843,39 @@ def main_loop(placeholders, val_placeholders, train_outs, train_summary_op,
+                # Write Mean_tower_loss summary
+                summary = tf.Summary.Value(tag='Mean_tower_loss_train',
+                                           simple_value=loss_value)
+                summary_str = tf.Summary(value=[summary])
```
What is the reason why this has to be done here rather than in the `train_summary_op`?

fvisin commented on main_loop_tf/validate.py:
Ditto

fvisin committed:
Improve and complete multi-GPU code

* Improve the structure by reusing code and using dictionaries

fvisin committed:
Fix print exception message in save_images

fral92 committed:
Merge branch 'multigpu' of https://github.com/fvisin/main_loop_tf into multigpu

fral92 commented on main_loop_tf/main.py:
```diff
@@ -259,37 +259,38 @@ def __run(build_model):
             inputs_per_gpu.append(tf.placeholder_with_default(
-                np.zeros(shape=[1]+cfg.input_shape[1:], dtype=cfg._FLOATX),
-                shape=cfg.input_shape))
+                np.zeros(shape=[cfg.batch_size] + cfg.input_shape[1:],
```
Is it not better to use 1 as default value for the batch_size? In the case we do not feed the placeholder the tower results are not taken so I think it is not important which is the default size, so why do not preserve memory?

fvisin replied:
I guess you are right. I was doing that to make sure it's a shape we know how to manage (in terms of number of GPUs), but I was mistaken: if we use the default we are already cutting out some GPUs, so there is no need for that. Can you change it back to 1 please?

fral92 commented on main_loop_tf/main.py:
```diff
@@ -324,53 +325,54 @@ def __run(build_model):
+        small_num_splits = {}
+        for s in ['train'] + ['eval_' + v for v in cfg.val_on_sets]:
+            params = cfg.dataset_params if s == 'train' else cfg.valid_params
+            this_set = cfg.Dataset(which_set=s, **params)
```
The problem here is that the "eval_" prefix does not allow the dataset loader to match some set. 
Could be passing s.split("_")[1] to which_set a possible solution?

fvisin replied:
The dataset might have an underscore in the name, in that case the train set would be wrongly split. Let's do this instead:
``` python
this_set = cfg.Dataset(which_set=s[5:] if s.startswith('eval_') else s, **params)
```

fral92 commented on main_loop_tf/main.py:
```diff
@@ -490,64 +494,59 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
         [inputs_per_gpu, labels_per_gpu] = placeholders
-
+        summaries = tf.get_collection_ref(
+            key='val_' + which_set + '_summaries')
```
If which_set already contains the 'eval_' prefix probably we could avoid 'val_' for the collection name.

fvisin replied:
Agreed. I intended to remove it as well, this one was left behind apparently! :+1:

fral92 commented on main_loop_tf/main.py:
```diff
@@ -719,50 +710,26 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                for el in summaries:
+                    if ('gpu%d' % i in el.name):
+                        this_summaries += [el]
+                    # TODO ??
```
Since the Mean_tower_loss summary is part of the shared graph after the towers, computing it for the 'full' and 'small' case will result in duplicated graph in Tensorboard (e.g., mtl_0, mtl_1). So my solution was to save the summary of the mean loss when the loss is computed after run in main loop so that we can give a unique name and the graph is updated each time.

fvisin replied:
Ok, now I understand why you did it this way. Let's talk about it in person. It's really ugly and prone to errors to save the summaries in the main loop rather than in the graph, there has to be another way.

fral92 commented on main_loop_tf/main.py:
```diff
@@ -719,50 +710,26 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+                    # if (i == 0 and
+                    #         'Mean_tower_loss_' + which_set in el.name):
+                    #     training_summaries += [el]
+            summary_op = tf.summary.merge(this_summaries)
```
The summary_op is always overwritten. 
I think that it is correct something like this summary_ops[case] = tf.summary.merge(this_summaries)

fvisin replied:
Right! We will have to change the main loop as well to select the right summary op.

fral92 commented on main_loop_tf/main.py:
```diff
@@ -593,89 +591,88 @@ def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
+            mask = tf.cast(tf.less_equal(labels, nclasses), tf.int32)
+        preds_flat = tf.reshape(preds, [-1])
+        m_iou, per_class_iou, cm_update_op, reset_cm_op = compute_mean_iou(
+            labels, preds_flat, nclasses, mask)
```
Is the confusion matrix Variable instantiated inside the compute_mean_iou the same? It could be that since in the case of smaller batch the for loop is performed twice the confusion matrix created are two. So the update_op and reset_op could be applied to two different cm.

fvisin replied:
As long as the tensors as argument are the same it should recover the same shared variable thanks to the variable scope:
see [here](https://github.com/tensorflow/tensorflow/blob/0b723590631432584d0761c03285eabb55116c6d/tensorflow/python/ops/metrics_impl.py#L909-L910) and the [definition of local variables](https://github.com/tensorflow/tensorflow/blob/0b723590631432584d0761c03285eabb55116c6d/tensorflow/python/ops/metrics_impl.py#L191-L196)

@marcociccone can you double check this please? If you are unsure we should verify this works as intended).

marcociccone replied:
I added the namescopes in order to distinguish the variables. These are the local variables with 2 gpus, just before calling the supervisor:

> [<tf.Variable 'train/total_confusion_matrix:0' shape=(2, 2) dtype=int64_ref>, <tf.Variable 'train_1/total_confusion_matrix:0' shape=(2, 2) dtype=int64_ref>, <tf.Variable 'eval_valid/total_confusion_matrix:0' shape=(2, 2) dtype=int64_ref>, <tf.Variable 'eval_valid_1/total_confusion_matrix:0' shape=(2, 2) dtype=int64_ref>]

fvisin committed:
Major graph refactor

* Select the parts of the graph we need, depending on the number of GPUs
  being used at each iteration (at runtime!)
* Select the summaries we care about, depending on the number of GPUs
  being used at each iteration (at runtime!)
* Create one graph per GPU and dataset split
* Collect all the summaries inside build_graph, as should be

fral92 committed:
Merge branch 'multigpu' of https://github.com/fvisin/main_loop_tf into multigpu
